modules/ocf_mirrors/files/collect-mirrors-stats

```diff
@@ -0,0 +1,168 @@
+    repeatedly for multiple projects and to avoid having to merge
+    dictionaries.
+    """
+
+    with fn.open('r') as f:
+        for n, line in enumerate(f, 1):
+            stats = line.split()
+
+            # apache log date format looks like [11/Jul/2017:00:05:16 -0700]
+            # line.split()[3][1:12] = 11/Jul/2017
+            # we need to dump dates that don't match because
+            # logrotate rotates the logs at 6am
+            try:
+                line_date = datetime.strptime(stats[3][1:12], '%d/%b/%Y')
+            except ValueError:
+                print('Error on line {}:\\  {}\\'.format(n, line))
+                traceback.print_exc()
+
+            if line_date.date() != log_date:
+                continue
+
+            # extract dist name from request url
+            # '/debian/pool/main/h/hwdata/...' -> 'debian'
+            dist = stats[6]
+            dist = dist.split('/')[1] if '/' in dist else dist
+
+            # record if we returned http 2xx/3xx
+            if stats[8][0] in ('2', '3') and dist in projects:
+                projects[dist]['up'] += int(stats[-2])
+                projects[dist]['down'] += int(stats[-1])
+            else:
+                projects['other']['up'] += int(stats[-2])
+                projects['other']['down'] += int(stats[-1])
+
+    return projects
+
+
+def process_rsync_log(filename):
+    """Parse Tx/Rx information out of an rsyncd log file.
+
+    This function is project-agnostic.
+    """
+
+    tx = 0
+    rx = 0
+    with filename.open('r') as f:
+        for line in f.readlines():
+            # log format is <datetime> <pid> sent <tx> bytes received <rx> bytes total size
+            if line.find('sent') > 0:
+                stats = line.split()
+                tx += int(stats[4])
+                rx += int(stats[7])
+
+    return tx, rx
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Process mirrors logs to calculate network usage '
+                                                 'and store in ocfstats')
+    parser.add_argument('-q', '--quiet', action='store_true',
+                        help='do not print stats after collecting them')
+    parser.add_argument('-n', '--dry-run', action='store_true',
+                        help='dry-run mode, do not write statistics to the database')
+    parser.add_argument('--apache', nargs='*',
+                        help='Apache log file(s) to process')
+    parser.add_argument('--rsync', nargs='*',
+                        help='rsync log files(s) to process')
+    parser.add_argument('date', nargs='?', default=date.today() - timedelta(1),
+                        help='date for use in filtering log entries')
+
+    args = parser.parse_args()
+
+    # first, figure out what projects we're getting stats for
+    sources = []
+    for project in MIRRORS_DATA_PATH.iterdir():
+        if project.is_dir() and not project.name.startswith('.'):
```

Comment:
err... `.` and `..` aren't going to be in the output of iterdir

Reply:
mm yeah holdover from the previous script